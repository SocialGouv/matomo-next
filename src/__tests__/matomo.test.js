import Router from "next/router";

import { init, push } from "../index";

let mockRouteChangeComplete;
jest.mock("next/router", () => ({
  events: {
    emit: (event, route) => {
      mockRouteChangeComplete(route);
      jest.fn();
    },
    on: (event, cb) => {
      mockRouteChangeComplete = cb;
    },
  },
  query: {},
}));

// default window.location.pathname
Object.defineProperty(window, "location", {
  value: {
    pathname: "/",
  },
});

describe("init", () => {
  beforeEach(() => {
    global._paq = [];
  });
  it("should create a js tag and initialize", () => {
    // we need to add a fake script node so
    // init can insert matomo tracker code before it
    document.head.appendChild(document.createElement("script"));
    init({ siteId: "42", url: "https://YO" });
    expect(global._paq).toMatchSnapshot();
  });
});

describe("push", () => {
  test("should append data to window._paq", () => {
    init({ siteId: "42", url: "YO" });
    window._paq = [];
    push(["trackEvent", "kikoo", "lol"]);
    expect(window._paq).toMatchSnapshot();
  });
});

describe("router.routeChangeComplete event", () => {
  test("should trackPageView with correct title on route change", () => {
    init({ siteId: "42", url: "YO" });
    window._paq = [];
    document.title = "test page";

    Router.events.emit("routeChangeComplete", "/path/to/hello?world");

    return new Promise((resolve) => {
      setTimeout(() => {
        expect(window._paq).toMatchSnapshot();
        resolve();
      }, 0);
    });
  });
  test("should use previousPath as referer on consecutive route change", () => {
    document.title = "test page 2";

    Router.events.emit("routeChangeComplete", "/path/to/hello2?world");

    return new Promise((resolve) => {
      setTimeout(() => {
        expect(window._paq).toMatchSnapshot();
        resolve();
      }, 0);
    });
  });
});

describe("excludeUrlsPatterns", () => {
  beforeEach(() => {
    global._paq = [];
  });
  it("should excluded login.php and token variables", () => {
    // we need to add a fake script node so
    // init can insert matomo tracker code before it
    document.head.appendChild(document.createElement("script"));
    init({
      excludeUrlsPatterns: [/^\/login.php/, /\?token=.+/],
      siteId: "42",
      url: "https://YO",
    });
    Router.events.emit("routeChangeComplete", "/login.php");
    Router.events.emit("routeChangeComplete", "/path/to/page.php");
    Router.events.emit("routeChangeComplete", "/path/to/page.php?token=pouet");
    return new Promise((resolve) => {
      setTimeout(() => {
        expect(window._paq).toMatchSnapshot();
        resolve();
      }, 0);
    });
  });
  it("should exclude initial page tracking", () => {
    window.location.pathname = "/change-password-pouet";
    document.head.appendChild(document.createElement("script"));
    init({
      excludeUrlsPatterns: [/^\/change-password/],
      siteId: "42",
      url: "https://YO",
    });

    return new Promise((resolve) => {
      setTimeout(() => {
        expect(window._paq).toMatchSnapshot();
        resolve();
      }, 0);
    });
  });

  it("should track initial page if not excluded", () => {
    window.location.pathname = "/some-page";
    document.head.appendChild(document.createElement("script"));
    init({
      excludeUrlsPatterns: [/^\/change-password/],
      siteId: "42",
      url: "https://YO",
    });

    return new Promise((resolve) => {
      setTimeout(() => {
        expect(window._paq).toMatchSnapshot();
        resolve();
      }, 0);
    });
  });
});

// todo: should track pageview on next router routeChangeComplete
