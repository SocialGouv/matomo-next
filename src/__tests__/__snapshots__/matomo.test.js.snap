// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`excludeUrlsPatterns should exclude initial page tracking 1`] = `
Array [
  Array [
    "enableLinkTracking",
  ],
  Array [
    "setTrackerUrl",
    "https://YO/matomo.php",
  ],
  Array [
    "setSiteId",
    "42",
  ],
]
`;

exports[`excludeUrlsPatterns should excluded login.php and token variables 1`] = `
Array [
  Array [
    "trackPageView",
  ],
  Array [
    "enableLinkTracking",
  ],
  Array [
    "setTrackerUrl",
    "https://YO/matomo.php",
  ],
  Array [
    "setSiteId",
    "42",
  ],
  Array [
    "setReferrerUrl",
    "/",
  ],
  Array [
    "setCustomUrl",
    "/path/to/page.php",
  ],
  Array [
    "setDocumentTitle",
    "test page 2",
  ],
  Array [
    "deleteCustomVariables",
    "page",
  ],
  Array [
    "setGenerationTimeMs",
    0,
  ],
  Array [
    "trackPageView",
  ],
]
`;

exports[`excludeUrlsPatterns should track initial page if not excluded 1`] = `
Array [
  Array [
    "trackPageView",
  ],
  Array [
    "enableLinkTracking",
  ],
  Array [
    "setTrackerUrl",
    "https://YO/matomo.php",
  ],
  Array [
    "setSiteId",
    "42",
  ],
]
`;

exports[`init should create a js tag and initialize 1`] = `
Array [
  Array [
    "trackPageView",
  ],
  Array [
    "enableLinkTracking",
  ],
  Array [
    "setTrackerUrl",
    "https://YO/matomo.php",
  ],
  Array [
    "setSiteId",
    "42",
  ],
]
`;

exports[`push should append data to window._paq 1`] = `
Array [
  Array [
    "trackEvent",
    "kikoo",
    "lol",
  ],
]
`;

exports[`router.routeChangeComplete event should trackPageView with correct title on route change 1`] = `
Array [
  Array [
    "setReferrerUrl",
    "/",
  ],
  Array [
    "setCustomUrl",
    "/path/to/hello",
  ],
  Array [
    "setDocumentTitle",
    "test page",
  ],
  Array [
    "deleteCustomVariables",
    "page",
  ],
  Array [
    "setGenerationTimeMs",
    0,
  ],
  Array [
    "trackPageView",
  ],
]
`;

exports[`router.routeChangeComplete event should use previousPath as referer on consecutive route change 1`] = `
Array [
  Array [
    "setReferrerUrl",
    "/",
  ],
  Array [
    "setCustomUrl",
    "/path/to/hello",
  ],
  Array [
    "setDocumentTitle",
    "test page",
  ],
  Array [
    "deleteCustomVariables",
    "page",
  ],
  Array [
    "setGenerationTimeMs",
    0,
  ],
  Array [
    "trackPageView",
  ],
  Array [
    "setReferrerUrl",
    "/path/to/hello",
  ],
  Array [
    "setCustomUrl",
    "/path/to/hello2",
  ],
  Array [
    "setDocumentTitle",
    "test page 2",
  ],
  Array [
    "deleteCustomVariables",
    "page",
  ],
  Array [
    "setGenerationTimeMs",
    0,
  ],
  Array [
    "trackPageView",
  ],
]
`;
